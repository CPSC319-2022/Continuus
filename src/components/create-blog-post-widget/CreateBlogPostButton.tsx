import { useRef, useState } from "react";
import { Spinner } from "../Spinner";

export const CreateBlogPostButton: React.FC<{ onClick?: () => void, unauthorized?: boolean, loading?: boolean }> = ({ onClick, unauthorized, loading }) => {
    const [hovered, setHovered] = useState(false);
    const disabledRoleClassNames = "bg-gray-400 cursor-not-allowed";
    const enabledRoleClassNames = "bg-green-400 hover:after:bg-green-100 hover:before:bg-green-100 hover:after:rotate-90 hover:before:-rotate-90 cursor-pointer";
    const tooltipRef = useRef<HTMLDivElement>(null);

    return (
        <>
            <div
                className="
                    fixed
                    top-[90%]
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2 
                    -translate-y-1/2
                "
            >
                {loading ? <Spinner size={4} /> :
                    <>
                        <div
                            className="
                            absolute
                            w-64
                            min-h-0
                            h-12
                            -translate-y-full
                            -translate-x-1/2
                            left-1/2
                            md:-translate-x-full
                            md:left-full
                            overflow-hidden
                            -top-2
                        "
                            style={{ height: tooltipRef.current?.scrollHeight || '16rem' }}
                        >
                            <div
                                className={`
                            ${(!unauthorized || !hovered) ? 'top-full' : 'top-0'}
                            absolute
                            w-full  
                            border-black
                            bg-gray-400
                            rounded-md
                            p-2
                            transition-all
                            after:absolute first-letter:
                            after:left-0
                            md:after:left-auto
                            after:right-0
                            after:my-0
                            after:mx-auto
                            after:w-0
                            after:h-0
                            after:border-t-[1rem]
                            after:border-solid
                            after:border-t-gray-400
                            after:border-l-[2rem]
                            after:border-r-[2rem]
                            after:border-x-transparent
                            after:transition-all
                        `}
                                ref={tooltipRef}
                            >
                                <p className="text-center md:text-right">You do not have authorization to create posts</p>
                            </div>
                        </div>
                        <button
                            onMouseEnter={() => setHovered(true)}
                            onMouseLeave={() => setHovered(false)}
                            onClick={() => !unauthorized && onClick?.()}
                            className={`
                        ${!unauthorized ? enabledRoleClassNames : disabledRoleClassNames} 
                        aspect-square 
                        w-16 
                        rounded-full 
                        after:block 
                        before:block 
                        after:bg-white 
                        before:bg-white 
                        after:absolute 
                        before:absolute 
                        after:top-1/2 
                        before:top-1/2 
                        after:left-1/2 
                        before:left-1/2 
                        after:-translate-x-1/2 
                        before:-translate-x-1/2 
                        after:-translate-y-1/2 
                        before:-translate-y-1/2 
                        after:h-8 
                        before:h-2 
                        after:w-2 
                        before:w-8 
                        after:transition-all 
                        before:transition-all 
                        `}
                        />
                    </>
                }
            </div>
        </>
    );
}