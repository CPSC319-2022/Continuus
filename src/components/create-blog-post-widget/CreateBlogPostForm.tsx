import Link from "next/link";
import { FormEvent, useState } from "react";

export interface CreateBlogPostFields {
    title: string;
    content: string;
}

export const CreateBlogPostForm: React.FC<{
    onSubmit: (fieldValues: CreateBlogPostFields) => void,
}> = ({
    onSubmit
}) => {
        const [title, setTitle] = useState("");
        const [content, setContent] = useState("");

        return (
            <form
                onSubmit={(event: FormEvent<HTMLFormElement>) => {
                    event.preventDefault();
                    onSubmit({
                        title,
                        content,
                    });
                }}
            >
                <div className="border-b border-b-gray-200 pl-4 pr-4 py-6 flex flex-row items-start">
                    <div>ProfileImage</div>
                    <div className="flex flex-col w-full">
                        <input
                            className="w-full bg-gray-100 rounded-md p-2 mb-4"
                            placeholder="Write title here"
                            type="text"
                            value={title}
                            onChange={(event) => setTitle(event.target.value)}
                        />
                        <textarea
                            className="w-full resize-none bg-gray-100 rounded-md h-96 p-2"
                            placeholder="Write your blog post here"
                            value={content}
                            onChange={(event) => setContent(event.target.value)}
                        />
                    </div>
                </div>
                <input type="submit" value="Create"
                    className="relative left-full -translate-x-full px-6 py-2 rounded-md bg-emerald-400 my-2 -ml-4 hover:text-white transition-colors cursor-pointer"
                />
            </form>
        );
    }