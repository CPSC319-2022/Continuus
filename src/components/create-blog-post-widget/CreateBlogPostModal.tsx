import { PropsWithChildren } from 'react';
import Modal from 'react-modal';

Modal.setAppElement("#__next")

export const CreateBlogPostModal: React.FC<PropsWithChildren<{
    isOpen: boolean,
    closeDisabled?: boolean,
    onRequestClose: () => void,
}>> = ({
    children,
    isOpen,
    closeDisabled,
    onRequestClose,
}) => {
        return (
            <Modal
                isOpen={isOpen}
                onRequestClose={() => closeDisabled || onRequestClose()}
                contentLabel="Create Blog Post"
                overlayClassName="fixed inset-0 z-20 bg-black/75"
                ariaHideApp={process.env.NODE_ENV !== 'test'}
                className="bg-white absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 w-1/2 rounded-md"
            >
                <div className="w-full h-12 flex flex-row justify-between items-center border-b border-b-gray-200 pl-4 pr-4">
                    <div className="font-semibold">Create Blog Post</div>
                    {
                        closeDisabled || <button
                            className="rounded-full after:content-['\2715'] hover:text-emerald-400 transition-colors"
                            onClick={() => closeDisabled || onRequestClose()}
                        />
                    }
                </div>
                {children}
            </Modal>
        );
    }