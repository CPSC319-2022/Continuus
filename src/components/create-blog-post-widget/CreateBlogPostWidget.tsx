import { CreateBlogPostButton } from "./CreateBlogPostButton";
import { CreateBlogPostForm } from "./CreateBlogPostForm";
import { BlogPostInputModal } from "../BlogPostInputModal";
import { useEffect, useState } from "react";
import { api } from "~/utils/api";
import { Spinner } from "../Spinner";
import { FiCheck } from "react-icons/fi";

export const CreateBlogPostWidget: React.FC = () => {
  const [isModalOpen, setModalOpen] = useState(false);
  const [isSuccessful, setSuccessful] = useState(false);

  const utils = api.useContext();

  const createBlogPostMutation = api.blogPost.create.useMutation({
    onSuccess() {
      return Promise.all([
        utils.blogPost.get.invalidate(),
        utils.blogPost.count.invalidate(),
      ]);
    },
  });
  const currUser = api.user.currentUser.useQuery();

  useEffect(() => {
    if (createBlogPostMutation.isSuccess) {
      setSuccessful(true);

      const timeoutId = setTimeout(() => {
        setModalOpen(false);
        setSuccessful(false);
      }, 2000);

      return () => {
        setSuccessful(false);
        clearTimeout(timeoutId);
      };
    }
  }, [createBlogPostMutation.isSuccess]);

  if (currUser.isLoading) {
    return <CreateBlogPostButton loading />;
  }

  if (
    currUser.data === null ||
    !!!currUser.data ||
    !["CONTRIBUTOR", "ADMIN"].includes(currUser.data.role)
  ) {
    return <CreateBlogPostButton unauthorized />;
  }

  const userId = currUser.data.id;

  return (
    <>
      <CreateBlogPostButton onClick={() => setModalOpen(true)} />
      <BlogPostInputModal
        action="Create"
        isOpen={isModalOpen}
        onRequestClose={() => {
          setModalOpen(false);
        }}
        closeDisabled={createBlogPostMutation.isLoading || isSuccessful}
      >
        {createBlogPostMutation.isLoading ? (
          <div className="flex h-96 w-full items-center justify-center">
            <Spinner size={12} />
          </div>
        ) : isSuccessful ? (
          <div className="flex h-96 w-full flex-col items-center justify-center">
            <FiCheck
              size={"12rem"}
              className="animate-pulse stroke-highlight-green"
            />
            <div className="text-lg">Successfully created the blog post</div>
          </div>
        ) : (
          <CreateBlogPostForm
            onSubmit={(fieldValues) =>
              createBlogPostMutation.mutate({
                data: {
                  title: fieldValues.title,
                  content: fieldValues.content,
                  userId: userId,
                },
              })
            }
          />
        )}
      </BlogPostInputModal>
    </>
  );
};
