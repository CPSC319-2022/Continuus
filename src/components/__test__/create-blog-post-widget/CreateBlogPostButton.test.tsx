import { describe, it, expect, beforeEach, vi } from 'vitest'
import { cleanup, render, screen } from "@testing-library/react";
import userEvent from '@testing-library/user-event'
import { CreateBlogPostButton } from '~/components/create-blog-post-widget/CreateBlogPostButton';


describe("Snapshot: CreateBlogPostButton", () => {

    beforeEach(() => {
        // this cleans up the 'screen'
        cleanup()
    })

    it("Loading Button", () => {
        render(<CreateBlogPostButton loading={true}/>);
        const createBlogPostButton = screen.getByLabelText("Create Blog Post Button");
        expect(createBlogPostButton).toMatchSnapshot();
    });

    it("Unauthorized Button", () => {
        // another way to take the snapshot of the UI component is to directly use the return value of 'render'
        // beware that this includes the container in the snapshot
        const createBlogPostButton = render(<CreateBlogPostButton unauthorized={true}/>).container;
        expect(createBlogPostButton).toMatchSnapshot();
    });

    it("Authorized Button", () => {
        render(<CreateBlogPostButton unauthorized={false}/>);
        const createBlogPostButton = screen.getByLabelText("Create Blog Post Button");
        expect(createBlogPostButton).toMatchSnapshot();
    });

    it("Hovering Unauthorized Button", async () => {
        const user = userEvent.setup()
        render(<CreateBlogPostButton unauthorized={true}/>);
        const createBlogPostButton = screen.getByLabelText("Create Blog Post Button");
        const actualButton = screen.getByRole("button");
        await user.hover(actualButton)
        expect(createBlogPostButton).toMatchSnapshot();
    });

    it("Hovering Authorized Button", async () => {
        const user = userEvent.setup()
        render(<CreateBlogPostButton unauthorized={false}/>);
        const createBlogPostButton = screen.getByLabelText("Create Blog Post Button");
        const actualButton = screen.getByRole("button");
        await user.hover(actualButton)
        expect(createBlogPostButton).toMatchSnapshot();
    });
    
    it("Clicking Authorized Button", async () => {
        const user = userEvent.setup()
        const onClick = vi.fn(() => {})
        render(<CreateBlogPostButton onClick={onClick} unauthorized={false}/>);
        const actualButton = screen.getByRole("button");
        await user.click(actualButton)

        expect(onClick).toHaveBeenCalledOnce();
    });
})