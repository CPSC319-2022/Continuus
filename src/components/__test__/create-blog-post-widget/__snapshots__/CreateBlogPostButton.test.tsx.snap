// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot: CreateBlogPostButton > Authorized Button 1`] = `
<div
  aria-label="Create Blog Post Button"
  class="
                    fixed
                    bottom-24
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2
                    translate-y-1/2
                "
>
  <div
    class="
                            absolute
                            w-64
                            min-h-0
                            h-12
                            -translate-y-full
                            -translate-x-1/2
                            left-1/2
                            md:-translate-x-full
                            md:left-full
                            overflow-hidden
                            -top-2
                        "
    style="height: 16rem;"
  >
    <div
      class="
                            top-full
                            absolute
                            w-full  
                            border-black
                            bg-gray-400
                            rounded-md
                            p-2
                            transition-all
                            after:absolute first-letter:
                            after:left-0
                            md:after:left-auto
                            after:right-0
                            after:my-0
                            after:mx-auto
                            after:w-0
                            after:h-0
                            after:border-t-[1rem]
                            after:border-solid
                            after:border-t-gray-400
                            after:border-l-[2rem]
                            after:border-r-[2rem]
                            after:border-x-transparent
                            after:transition-all
                        "
    >
      <p
        class="p-3 text-center md:text-left"
      >
        You do not have authorization to create posts
      </p>
    </div>
  </div>
  <button
    class="
                        bg-highlight-green hover:bg-gray-700 hover:scale-110 bg-highlight-green cursor-pointer transition-all 
                        aspect-square
                        w-16 
                        rounded-full 
                        after:block 
                        before:block 
                        after:bg-white 
                        before:bg-white
                        after:absolute 
                        before:absolute
                        after:top-1/2 
                        before:top-1/2 
                        after:left-1/2 
                        before:left-1/2 
                        after:-translate-x-1/2 
                        before:-translate-x-1/2 
                        after:-translate-y-1/2 
                        before:-translate-y-1/2 
                        after:h-8 
                        before:h-2 
                        after:w-2 
                        before:w-8 
                        after:transition-all 
                        before:transition-all 
                        "
  />
</div>
`;

exports[`Snapshot: CreateBlogPostButton > Hovering Authorized Button 1`] = `
<div
  aria-label="Create Blog Post Button"
  class="
                    fixed
                    bottom-24
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2
                    translate-y-1/2
                "
>
  <div
    class="
                            absolute
                            w-64
                            min-h-0
                            h-12
                            -translate-y-full
                            -translate-x-1/2
                            left-1/2
                            md:-translate-x-full
                            md:left-full
                            overflow-hidden
                            -top-2
                        "
    style="height: 16rem;"
  >
    <div
      class="
                            top-full
                            absolute
                            w-full  
                            border-black
                            bg-gray-400
                            rounded-md
                            p-2
                            transition-all
                            after:absolute first-letter:
                            after:left-0
                            md:after:left-auto
                            after:right-0
                            after:my-0
                            after:mx-auto
                            after:w-0
                            after:h-0
                            after:border-t-[1rem]
                            after:border-solid
                            after:border-t-gray-400
                            after:border-l-[2rem]
                            after:border-r-[2rem]
                            after:border-x-transparent
                            after:transition-all
                        "
    >
      <p
        class="p-3 text-center md:text-left"
      >
        You do not have authorization to create posts
      </p>
    </div>
  </div>
  <button
    class="
                        bg-highlight-green hover:bg-gray-700 hover:scale-110 bg-highlight-green cursor-pointer transition-all 
                        aspect-square
                        w-16 
                        rounded-full 
                        after:block 
                        before:block 
                        after:bg-white 
                        before:bg-white
                        after:absolute 
                        before:absolute
                        after:top-1/2 
                        before:top-1/2 
                        after:left-1/2 
                        before:left-1/2 
                        after:-translate-x-1/2 
                        before:-translate-x-1/2 
                        after:-translate-y-1/2 
                        before:-translate-y-1/2 
                        after:h-8 
                        before:h-2 
                        after:w-2 
                        before:w-8 
                        after:transition-all 
                        before:transition-all 
                        "
  />
</div>
`;

exports[`Snapshot: CreateBlogPostButton > Hovering Unauthorized Button 1`] = `
<div
  aria-label="Create Blog Post Button"
  class="
                    fixed
                    bottom-24
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2
                    translate-y-1/2
                "
>
  <div
    class="
                            absolute
                            w-64
                            min-h-0
                            h-12
                            -translate-y-full
                            -translate-x-1/2
                            left-1/2
                            md:-translate-x-full
                            md:left-full
                            overflow-hidden
                            -top-2
                        "
    style="height: 16rem;"
  >
    <div
      class="
                            top-0
                            absolute
                            w-full  
                            border-black
                            bg-gray-400
                            rounded-md
                            p-2
                            transition-all
                            after:absolute first-letter:
                            after:left-0
                            md:after:left-auto
                            after:right-0
                            after:my-0
                            after:mx-auto
                            after:w-0
                            after:h-0
                            after:border-t-[1rem]
                            after:border-solid
                            after:border-t-gray-400
                            after:border-l-[2rem]
                            after:border-r-[2rem]
                            after:border-x-transparent
                            after:transition-all
                        "
    >
      <p
        class="p-3 text-center md:text-left"
      >
        You do not have authorization to create posts
      </p>
    </div>
  </div>
  <button
    class="
                        bg-gray-200 cursor-not-allowed 
                        aspect-square
                        w-16 
                        rounded-full 
                        after:block 
                        before:block 
                        after:bg-white 
                        before:bg-white
                        after:absolute 
                        before:absolute
                        after:top-1/2 
                        before:top-1/2 
                        after:left-1/2 
                        before:left-1/2 
                        after:-translate-x-1/2 
                        before:-translate-x-1/2 
                        after:-translate-y-1/2 
                        before:-translate-y-1/2 
                        after:h-8 
                        before:h-2 
                        after:w-2 
                        before:w-8 
                        after:transition-all 
                        before:transition-all 
                        "
  />
</div>
`;

exports[`Snapshot: CreateBlogPostButton > Loading Button 1`] = `
<div
  aria-label="Create Blog Post Button"
  class="
                    fixed
                    bottom-24
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2
                    translate-y-1/2
                "
>
  <div
    class="inline-block animate-spin rounded-full border-solid border-current border-r-transparent align-[-0.125em] text-highlight-green motion-reduce:animate-[spin_1.5s_linear_infinite]"
    role="status"
    style="width: 4rem; height: 4rem; border-width: 0.3333333333333333rem;"
  >
    <span
      class="absolute -m-px h-px w-px overflow-hidden whitespace-nowrap border-0 p-0 [clip:rect(0,0,0,0)]"
    />
  </div>
</div>
`;

exports[`Snapshot: CreateBlogPostButton > Unauthorized Button 1`] = `
<div>
  <div
    aria-label="Create Blog Post Button"
    class="
                    fixed
                    bottom-24
                    left-1/2 
                    z-10
                    md:left-[95%] 
                    -translate-x-1/2
                    translate-y-1/2
                "
  >
    <div
      class="
                            absolute
                            w-64
                            min-h-0
                            h-12
                            -translate-y-full
                            -translate-x-1/2
                            left-1/2
                            md:-translate-x-full
                            md:left-full
                            overflow-hidden
                            -top-2
                        "
      style="height: 16rem;"
    >
      <div
        class="
                            top-full
                            absolute
                            w-full  
                            border-black
                            bg-gray-400
                            rounded-md
                            p-2
                            transition-all
                            after:absolute first-letter:
                            after:left-0
                            md:after:left-auto
                            after:right-0
                            after:my-0
                            after:mx-auto
                            after:w-0
                            after:h-0
                            after:border-t-[1rem]
                            after:border-solid
                            after:border-t-gray-400
                            after:border-l-[2rem]
                            after:border-r-[2rem]
                            after:border-x-transparent
                            after:transition-all
                        "
      >
        <p
          class="p-3 text-center md:text-left"
        >
          You do not have authorization to create posts
        </p>
      </div>
    </div>
    <button
      class="
                        bg-gray-200 cursor-not-allowed 
                        aspect-square
                        w-16 
                        rounded-full 
                        after:block 
                        before:block 
                        after:bg-white 
                        before:bg-white
                        after:absolute 
                        before:absolute
                        after:top-1/2 
                        before:top-1/2 
                        after:left-1/2 
                        before:left-1/2 
                        after:-translate-x-1/2 
                        before:-translate-x-1/2 
                        after:-translate-y-1/2 
                        before:-translate-y-1/2 
                        after:h-8 
                        before:h-2 
                        after:w-2 
                        before:w-8 
                        after:transition-all 
                        before:transition-all 
                        "
    />
  </div>
</div>
`;
